### PR TITLE
remove old diagnostics for files without new diagnostic messages

### PR DESCRIPTION
**Pull request by AlexHaxe**
Once you reach 100% coverage for a file, old diagnostics would stay until you modified it.

### CODE REVIEW

**Comment by AlexHaxe** [comment]
My status bar wasn't working properly (it would only show `Coverage: …` once after opening a project, but would vanish and not come back when opening a new tab.
It seems `onDidCloseTextDocument` gets called after `onDidOpenTextDocument` or `onDidChangeTextDocument` so it would hide any status set by those.

Also I get two calls to `showStatus` one with a valid file name, and one with the same file name with `.git` appended to it (maybe it's one of my extensions), so when clearing status text I had to check file name against `activeTextEditor`.

File: src/extension.ts [hunk 1]
```diff
@@ -41,7 +41,9 @@ export function activate(context: ExtensionContext) {
     }
   });
   workspace.onDidCloseTextDocument(() => {
-    statusBar.hide();
+    if (!window.activeTextEditor) {
```
Comment (markis) [inline src/extension.ts]: What is this doing? `activeTextEditor` and `statusBar` seem unrelated.

File: src/extension.ts [hunk 2]
```diff
@@ -74,6 +76,11 @@ export function activate(context: ExtensionContext) {
         statusBar.text = `Coverage: ${lines.hit}/${lines.found} lines`;
         statusBar.show();
       }
+    } else {
+      const activeTextEditor = window.activeTextEditor;
+      if (activeTextEditor && (activeTextEditor.document.uri.fsPath.toLowerCase() === file)) {
```
Comment (markis) [inline src/extension.ts]: If it is possible for the `activeTextEditor.document.uri.fsPath` to not equal the `file`, then maybe we shouldn't pass file in and maybe should just use `activeTextEditor.document.uri.fsPath`?  Also, seems like we should do this for entire function.

File: src/extension.ts [hunk 3]
```diff
@@ -96,6 +103,8 @@ export function activate(context: ExtensionContext) {
 
         if (diagnosticsForFiles.length > 0) {
           diagnostics.set(Uri.file(coverage.file), diagnosticsForFiles);
+        } else {
+          diagnostics.delete(Uri.file(coverage.file));
```
Comment (markis) [inline src/extension.ts]: This makes a lot of sense! 👍

File: src/extension.ts (hunk 1 above)
Comment (AlexHaxe) [reply src/extension.ts]: It hides `statusBar` only when there is no active editor. 
I had cases where `onDidClose` came after a `onDidOpen` / `onDidChange`, so `statusBar` was hidden even though the then active editor had some coverage.

File: src/extension.ts (hunk 2 above)
Comment (AlexHaxe) [reply src/extension.ts]: ok, refactored..

Should we also rewrite `onDidCloseTextDocument` to juse call `showStatus` and move `statusBar.hide()` into there?

**Comment by AlexHaxe** [comment]
ok, I think I'm done.

It's now a bit cleaner, `showStatus` manages anything status bar related. And in my tests it always shows a status bar for files with coverage and no status bar for all files without coverage.
